### PR TITLE
[JENKINS-34369] protect against a NPE form GitSCM.getBuildData

### PR DESCRIPTION
**Pull request by kuisathaverat**
A NPE is not checked into PreBuildMerge.decorateRevisionToBuild

@reviewbybees

### CODE REVIEW

**Comment by KostyaSha** [comment]
https://issues.jenkins-ci.org/browse/JENKINS-34369

**Comment by KostyaSha** [comment]
No clear description, no steps to reproduce, not test case 👎

**Comment by kuisathaverat** [comment]
Method GitSCM.getBuildData could return null

`public @CheckForNull BuildData getBuildData(Run build) {
`

but in PreBuildMerge.decorateRevisionToBuild does not checked if a null is returned

```
public Revision decorateRevisionToBuild(GitSCM scm, Run<?, ?> build, GitClient git, TaskListener listener, Revision marked, Revision rev) throws IOException, InterruptedException {
...
    scm.getBuildData(build).saveBuild(new Build(marked,rev, build.getNumber(), FAILURE));
           
    throw new AbortException("Branch not suitable for integration as it does not merge cleanly: " + ex.getMessage());
...
```

**Comment by kuisathaverat** [comment]
@MarkEWaite @jglick @mlgiroux @ndeloof 
Please could you review this pull request and express your opinion? about protect this method against a annotated method with @CheckForNull, it is simple, safe and does not hide the real exception.

**Comment by mrdfuse** [comment]
@MarkEWaite @jglick @mlgiroux @ndeloof 
any update on this pr?

File: src/main/java/hudson/plugins/git/extensions/impl/PreBuildMerge.java [hunk 1]
```diff
@@ -85,7 +86,10 @@ public Revision decorateRevisionToBuild(GitSCM scm, Run<?, ?> build, GitClient g
             // record the fact that we've tried building 'rev' and it failed, or else
             // BuildChooser in future builds will pick up this same 'rev' again and we'll see the exact same merge failure
             // all over again.
-            scm.getBuildData(build).saveBuild(new Build(marked,rev, build.getNumber(), FAILURE));
+            BuildData bd = scm.getBuildData(build);
+            if(bd != null){
+                bd.saveBuild(new Build(marked,rev, build.getNumber(), FAILURE));
+            }
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/git/extensions/impl/PreBuildMerge.java]: If build data is missing, maybe makes sense to create it or to at least log the warning about unexpected behavior

**Comment by KostyaSha** [comment]
> @KostyaSha
> 
> Method GitSCM.getBuildData could return null

This is not the reason, this is how code works. For understanding fix everybody must know real root reason and steps to reproduce. Referenced jira issue contains cases, but they are not from topic starter.

**Comment by aheritier** [comment]
@kuisathaverat what is the status here ? can we move forward ?

**Comment by kuisathaverat** [comment]
@aheritier I added the suggestion from @oleg-nenashev, the steps to replicate the issue are in the Jira, I think that I can not push it more

**Comment by kuisathaverat** [comment]
@reviewbybees Could we close the review and move forward?

**Comment by mrooney** [comment]
Running into this issue in production, would love a release! JIRA issue seems fairly detailed. Anything I can do to help?

**Comment by MarkEWaite** [comment]
@mcrooney, you could compile the code and confirm it resolves the issue you're seeing, then report that to this list.

**Comment by stephenc** [comment]
:bee:

**Comment by oleg-nenashev** [comment]
re-:bee:

**Comment by kuisathaverat** [comment]
@reviewbybees done

**Pull request merged**